ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/endpoint/RetryingServiceEndpointFinder.java

```diff
@@ -0,0 +1,129 @@
+            throws ServiceEndpointLookupException, InterruptedException {
+
+        String targetInstanceId = serviceEndpointRequest.getTargetInstance().getTargetInstanceId();
+
+        Retryer<T> loop = buildRetryer(serviceEndpointRequest);
+
+        T serviceEndpoint;
+        try {
+            serviceEndpoint = loop.call(() -> delegate.getServiceEndpoint(serviceEndpointRequest));
+        } catch (ExecutionException e) {
+            try {
+                throw e.getCause();
+            } catch (RuntimeException | Error | InterruptedException cause) {
+                throw cause;
+            } catch (Throwable cause) {
+                throw new IllegalStateException(cause);
+            }
+        } catch (RetryException e) {
+            if (Thread.interrupted()) {
+                throw new InterruptedException(String.format(
+                        "Interrupted while discovering service endpoint for instance %s", targetInstanceId));
+            }
+
+            // All attempts failed
+            LOG.info("Failed to discover service endpoint for instance {}", targetInstanceId);
+            Attempt<?> lastFailedAttempt = e.getLastFailedAttempt();
+            if (lastFailedAttempt.hasException()) {
+                try {
+                    throw lastFailedAttempt.getExceptionCause();
+                } catch (RuntimeException | Error | InterruptedException | ServiceEndpointLookupException cause) {
+                    throw cause;
+                } catch (Throwable cause) {
+                    throw new IllegalStateException(cause);
+                }
+            } else {
+                throw new ServiceEndpointLookupException(String.format(
+                        "Timed out while discovering service endpoint for instance %s", targetInstanceId), true);
+            }
+        }
+        return serviceEndpoint;
+    }
+
+    /**
+     * Builds a retryer for the specified request.
+     *
+     * @param serviceEndpointRequest the service endpoint request
+     * @param <T> the type of service endpoint
+     * @return the retryer
+     */
+    private <T extends ServiceEndpoint> Retryer<T> buildRetryer(ServiceEndpointRequest<T> serviceEndpointRequest) {
+        String targetInstanceId = serviceEndpointRequest.getTargetInstance().getTargetInstanceId();
+        Optional<ZonedDateTime> waitUntilTime = serviceEndpointRequest.getWaitUntilTime();
+
+        StopStrategy stop;
+        if (waitUntilTime.isPresent()) {
+            // Given the time at which waiting should stop,
+            // get the available number of millis from this instant
+            ZonedDateTime timeout = waitUntilTime.get();
+            stop = StopStrategies2.waitUntilDateTime(timeout);
+            LOG.info("Trying until {} to discover service endpoint for instance {}", timeout, targetInstanceId);
+        } else {
+            stop = StopStrategies.neverStop();
```

Comment:
More of a side comment, but: Any way to avoid having an unbounded wait?

Reply:
I changed the logging to a warn. Once all the configuration stuff is wired up, I will consider throwing an exception here if it's not configured.